Introduce complexity analysis phase

fishcakez opened the pull request:
By default adds 1 per field in the query. This can be overriden by
using a custom complexity analyser that is 2-arity function. The first
argument is the field's context and the second argument is the
complexity of the children.

Heavily influnced by Sangria.

fishcakez commented on lib/absinthe/phase/document/complexity.ex:
```diff
@@ -0,0 +1,58 @@
+
+  use Absinthe.Phase
+
+  @default_complexity 1
```
1 doesnt seem good default if using integers because the only number less than 1 is 0, and the next number is double. should it be floats? Or default to bigger number?

benwilson512 replied:
The default resolution function is essentially just a Map.get, and it is for all intents and purposes impossible to resolve any faster than that. Thus I think a default complexity of 1 isn't so bad, since it doesn't seem feasible to go lower anyway.

benwilson512 replied:
This does bring up a good general question about how to weight this stuff. A field that hits the database might have a reasonable default complexity of 100. I wonder how sangria weights stuff.

fishcakez replied:
Thats a very good point about `Map.get`

jparise replied:
Related to this, what should 0 represent (as a value returned by a complexity function)? Is there something "less" complex than `Map.get`? Would that be something that just returns a literal / constant value?

benwilson512 replied:
Returning a literal is slightly faster than doing a Map.get but honestly both a literal and a Map.get are fast enough that the resolution complexity is less than the overhead of resolving the field in general. I feel like 1 is a good floor, I can't think of any reason to go less.

fishcakez replied:
I have allowed `0` because a field with no children will have a child complexity of 0 passed to it.

fishcakez commented on lib/absinthe/phase/document/complexity.ex:
```diff
@@ -0,0 +1,58 @@
+  def run(input, options \\ []) do
+    %{operations: ops} = result = Blueprint.postwalk(input, &handle_node/1)
+    complexity = sum_complexity(ops)
+    max = Keyword.get(options, :max_complexity, :infinity)
```
Please help picking default max - if any

fishcakez commented on lib/absinthe/phase/document/complexity.ex:
```diff
@@ -0,0 +1,58 @@
+  defp field_complexity(%{complexity: {mod, fun}}, arg, child_complexity) do
+    apply(mod, fun, [arg, child_complexity])
+  end
+
```
Should add nice error?

benwilson512 replied:
Not entirely sure what this is referring to.

fishcakez replied:
Instead of FunctionClauseError, the resolve fun has a nice error IIRC. The resolver also does some other things with the function but I'm not sure what they are for.

benwilson512 replied:
I would prefer to handle this as either a schema validation or at the call sight of the `complexity` macro. Then in this actual phase at runtime we can leave it as is.

fishcakez commented on lib/absinthe/pipeline.ex:
```diff
@@ -92,6 +92,7 @@ defmodule Absinthe.Pipeline do
       # Prepare for Execution
       Phase.Document.CascadeInvalid,
       Phase.Document.Flatten,
+      {Phase.Document.Complexity, options},
```
Should this be optional phase? I think current implementation *must* go in this position.

benwilson512 replied:
Good question. If it's non optional I think a relatively high default should be picked. We also need to make sure to propagate the ability to set this option to things like Absinthe.Plug.

I'm inclined to include it by default, it's relatively easy to omit a phase from a pipeline if desired.

bruce replied:
Can we just support a `nil` default, making the phase "installed" by default, but essentially no-op?

jparise replied:
The other consideration is the cost of running the complexity analysis itself.

benwilson512 replied:
Well, the cost of doing the analysis is very low. It's basically just walking the tree and building an accumulator. It's certainly no worse than any other phase.

fishcakez replied:
Still always running the analysis but doing nothing with it. Thoughts?

fishcakez commented on test/lib/absinthe/phase/document/complexity_test.exs:
```diff
@@ -0,0 +1,88 @@
+defmodule Absinthe.Phase.Document.Arguments.ComplexityTest do
```
typo in module name oops, will fix on review

benwilson512 commented on lib/absinthe/blueprint/document/operation.ex:
```diff
@@ -18,6 +18,7 @@ defmodule Absinthe.Blueprint.Document.Operation do
     # Populated by phases
     flags: %{},
     schema_node: nil,
+    complexity: nil,
```
Would it make sense to initialize this to the value of the default complexity?

Doing so would simplify a couple of function clauses in the phase.

fishcakez replied:
I didn't like to set the complexity in case someone wanted to make use of the complexity in another phase.

benwilson512 replied:
Makes sense.

benwilson512 commented on lib/absinthe/phase/document/complexity.ex:
```diff
@@ -0,0 +1,58 @@
+  """
+  @spec run(Blueprint.t, Keyword.t) :: Phase.result_t
+  def run(input, options \\ []) do
+    %{operations: ops} = result = Blueprint.postwalk(input, &handle_node/1)
```
This is a great use of postwalk!

benwilson512 replied:
@fishcakez Have you considered using the 2 arity postwalk by the way? I do like having each node tagged with its aggregate weight, but if you used 2 arity version the tree would remain unchanged and you'd still get a total.

fishcakez replied:
My initial version did that but I thought complexity values might be useful in later phases. See https://github.com/absinthe-graphql/absinthe/pull/260#discussion_r99453556

benwilson512 commented on lib/absinthe/phase/document/complexity.ex:
```diff
@@ -0,0 +1,58 @@
+  end
+  defp field_complexity(%{complexity: complexity}, arg, child_complexity)
+       when is_function(complexity, 2) do
+    complexity.(arg, child_complexity)
```
`arg` here is really all the `args` to the function right? Do we want to pass in just the arguments or the entire field blueprint node? I'm not entirely sure what would be needed additionally out of the blueprint node other than the argument data but I'm not sure we want to limit ourselves.

fishcakez replied:
Unsure, I haven't played with the implementation enough. I considered adding an arity 3 version with more information.

bruce commented on lib/absinthe/phase/document/complexity.ex:
```diff
@@ -0,0 +1,58 @@
+    complexity = sum_complexity(ops)
+    max = Keyword.get(options, :max_complexity, :infinity)
+    if complexity > max do
+      {:error, "complexity is #{complexity}, which is above maximum #{max}"}
```
It might be nice for us to generate an error with more metadata here, too, e.g.:

```
{:error, message: "complexity of 6 exceeds limit of 4", complexity: [calculated: 6, limit: 4]}
```

Actual message and metadata up for discussion -- I'd just like us to include more data for automated systems, generally, in addition to textual error messages when we can.

jparise replied:
I also wonder how we can best help the client react to this error. For example, should there be some more specific context about the "heaviest" part of the query?

bruce replied:
Excellent point. Perhaps a by-weight ordered list? A top 5? Hmmm...

fishcakez replied:
Should we split the phase in two? The first phase to analyse complexity and the second to handle it? I agree that it is opaque at present.

bruce commented on lib/absinthe/schema/notation.ex:
```diff
@@ -423,6 +423,21 @@ defmodule Absinthe.Schema.Notation do
   end
 
+  @placement {:complexity, [under: [:field]]}
+  defmacro complexity(func_ast) do
```
I am always impressed when someone manages to modify our insane Notation macros. Bravo 😉

jparise commented on lib/absinthe/type/field.ex:
```diff
@@ -111,6 +120,37 @@ defmodule Absinthe.Type.Field do
+  ### Custom Complexity
+
+  When accepting arguments, however, you probably need do use them for
+  something. Here's an example of definining a field that looks up a at most
```
I think the ` a ` might be unintentional here?

fishcakez replied:
Thanks

bruce commented:
@fishcakez With regard to complexity functions having access to execution environment information, some scenarios...

Forgive me if some of these feel a bit contrived, but I've been consistently surprised with the ways people want to modify execution behavior, and access to execution environment information (the schema, the field definition, the context, the root value, etc) seems to be a common need that enables users. So far, I think the "just give them the current execution environment information so they can do whatever" has served us well.  One example: just recently, giving custom scalars' `parse` functions access to the context helped us build the `:upload` type that's now part of absinthe_plug.

Anyhow, some random ideas that just may be kooky enough for someone to want:

- A field that needs to load data from different sources (with differing performance characteristics) depending on whether the request is authenticated/authorized, and so is therefore less/more "complex" (or costly) based on the API user.
- Some system status (services down, caches up, load spiking, etc) that changes how fields needs to access data (and can be stored in the context once at the beginning of the request to then be retrieved by fields) needs to be factored into their complexity analysis.
- It would be handy to use a single complexity function that could be aware of what field it's being called for, and can change the way it's calculating dependency based on the field definition. 

I've actually built this context-aware capability into another branch, `complexity-context` -- you can see a working test case here: https://github.com/absinthe-graphql/absinthe/blob/complexity-info/test/lib/absinthe/phase/document/complexity_test.exs#L70

bruce commented:
BTW, I'm not advocating for dynamic complexity definitions -- but I don't think we should arbitrarily stop them from being possible.

fishcakez commented:
I have updated the PR based on feedback:

- Support passing more context to complexity analysers
- Support integer complexity instead of analyser function
- Split complexity analysis and complexity result handling (allows custom phases to alter or explain complexity)
- Add detailed errors
- Only run complexity analysis on current operation

benwilson512 commented on lib/absinthe/type/field.ex:
```diff
@@ -119,10 +169,12 @@ defmodule Absinthe.Type.Field do
                default_value: any,
                args: %{(binary | atom) => Absinthe.Type.Argument.t} | nil,
                resolve: resolver_t | nil,
+               complexity: complexity_t | nil,
```
This should either include `non_neg_integer`, or `complexity_t` should.

fishcakez replied:
Fixed

benwilson512 commented on lib/absinthe/phase/document/complexity/analysis.ex:
```diff
@@ -0,0 +1,76 @@
+  end
+  defp field_complexity(%{complexity: {mod, fun}}, arg, child_complexity, info_data, _) do
+    info = struct(Complexity, Map.put(info_data, :definition, node))
+    apply(mod, fun, [arg, child_complexity, info])
```
Should we `max(result_of_function, 0)` this and the other complexity function invocations? I normally wouldn't code so defensively but I'm concerned that someone who does their logic badly may end up with weird errors.

fishcakez replied:
Is it better to error?

benwilson512 replied:
An error feels a bit severe, maybe a `Logger.warn` ?

benwilson512 replied:
Example situation where `max` may be useful: `{foo(limit: -1000) {y], x { ...ThisCanBeArbitrarilyComplexNow}`. If the logic for computing the complexity of `foo` is not careful to restrict the values of `limit` the `foo` field would have substantial negative complexity, which gives wide leeway for the rest of the document to be as complex as you want.

While obviously complexity functions ought to take precautions, I feel like preventing this kind of attack is within the purview of what this phase should do.

fishcakez replied:
I agree we need to prevent that but does setting to 0 prevent it?

bruce replied:
I think it would be better to error if negative complexity is returned for a field, similar to what we do when, eg, there is a non_null constraint on a field when a null is provided, or a bad resolver result. Warning and/or silently setting to 0. I think we _should_ be harsh.

benwilson512 replied:
Yeah that's fair, maxing to 0 only mitigates the issue rather than preventing it.

fishcakez replied:
I've added a pretty error for this (it will also catch bad complexity analysers) and it uncovered a bug. The branch is walking fields and selections which means we walk over children multiple times.

jparise commented on lib/absinthe/type/field.ex:
````diff
@@ -111,6 +121,47 @@ defmodule Absinthe.Type.Field do
+  end
+  ```
+
+  Custom complexity functions are passed two arguments:
````
Update this documentation to also include the 3-arity version of the complexity function?

fishcakez replied:
Updated, thank you.

benwilson512 commented:
I'd prefer to solve this more generically as this isn't the first such
phase to have this issue.

One option is to set a flag on the node in the flatten phase, which is used
by the prewalk code to choose which keys to walk

On Sun, Feb 5, 2017 at 10:57 AM James Fish <notifications@github.com> wrote:

> *@fishcakez* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In lib/absinthe/phase/document/complexity/analysis.ex
> <https://github.com/absinthe-graphql/absinthe/pull/260>:
>
>
> > +  defp field_complexity(%{complexity: complexity}, _, _, _, _)
>
> +       when is_integer(complexity) and complexity >= 0 do
>
> +    complexity
>
> +  end
>
> +  defp field_complexity(%{complexity: complexity}, arg, child_complexity, _, _)
>
> +       when is_function(complexity, 2) do
>
> +    complexity.(arg, child_complexity)
>
> +  end
>
> +  defp field_complexity(%{complexity: complexity}, arg, child_complexity, info_data, node)
>
> +       when is_function(complexity, 3) do
>
> +    info = struct(Complexity, Map.put(info_data, :definition, node))
>
> +    complexity.(arg, child_complexity, info)
>
> +  end
>
> +  defp field_complexity(%{complexity: {mod, fun}}, arg, child_complexity, info_data, _) do
>
> +    info = struct(Complexity, Map.put(info_data, :definition, node))
>
> +    apply(mod, fun, [arg, child_complexity, info])
>
>
>
> I believe need to walk "manually" through the tree only touching fields
> but please confirm.
>
>
>
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/absinthe-graphql/absinthe/pull/260>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAPICRiT4ZquWN3S21EUZ71YnC0GK6YPks5rZfFagaJpZM4L272S>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

fishcakez commented:
I've added a flag as suggested. Also disabled the complexity analyse by default, and it is enabled by an option but it is not clear where to document the phase specific options.

benwilson512 commented:
Thank you!
